infrastructure/instances.tf

```diff
@@ -16,21 +16,86 @@ data "aws_ami" "ubuntu" {
   }
 }
 
-resource "aws_instance" "data_refinery_worker_1" {
+# We need to transfer the Nomad Job Specifications onto the instance
+# so we can register them. However we use these job specifications to
+# store environment variables, so they are on Github as templates. To
+# format them with the production environment variables, we need to
+# run the following command. By doing so with a `null_resource` we're
+# able to avoid having a prerequisite step to running `terraform
+# apply`.
+resource "null_resource" "format-nomad-job-specs" {
+  provisioner "local-exec" {
+    command = "cd .. && ./workers/format_nomad_with_env.sh -e prod -o $(pwd)/infrastructure/nomad-job-specs/"
+  }
+}
+
+# This script installs Nomad.
+data "local_file" "install-nomad-script" {
+  filename = "../install_nomad.sh"
+}
+
+# This is the configuration for the Nomad Server.
+data "local_file" "nomad-server-config" {
+  filename = "nomad-configuration/server.hcl"
+}
+
+# This is a Nomad Job Specification file built by ${null_resource.format-nomad-job-specs}.
+data "local_file" "downloader-job-spec" {
+  filename = "nomad-job-specs/downloader.nomad"
+}
+
+# This is another Nomad Job Specification file built by ${null_resource.format-nomad-job-specs}.
+data "local_file" "processor-job-spec" {
+  filename = "nomad-job-specs/processor.nomad"
+}
+
+# This script smusher exists in order to be able to circumvent a
+# limitation of AWS which is that you get one script and one script
+# only to set up the instance when it boots up. Because there is only
+# one script you cannot place additional files your script may need
+# onto the instance. Therefore this script smusher templates the files
+# the instance-user-data.sh script needs into it, so that once it
+# makes its way onto the instance it can spit them back out onto the
+# disk.
+data "template_file" "nomad-server-script-smusher" {
+  template = "${file("nomad-configuration/nomad-server-instance-user-data.tpl.sh")}"
+
+  vars {
+    downloader_job_spec = "${data.local_file.downloader-job-spec.content}"
+    processor_job_spec = "${data.local_file.processor-job-spec.content}"
+    install_nomad_script = "${data.local_file.install-nomad-script.content}"
+    nomad_server_config = "${data.local_file.nomad-server-config.content}"
+  }
+}
+
+# This instance will run the Nomad Server. We will want three of
+# these, but one server needs to start first so that the others can be
+# aware of its IP address to join it.
+# Those other servers do not yet exist in the current configuration.
+# TODO: create those other servers!
+resource "aws_instance" "nomad-server-1" {
   ami = "${data.aws_ami.ubuntu.id}"
   instance_type = "t2.xlarge"
```

Comment:
Is this instance type overkill for what it is used for? Still not super expensive but I don't know what the requirements are for these machines, it shouldn't be too huge.

Reply:
Oh yeah that's a really good point. I think this will be an area where we'll need to do tuning. [Nomad says](https://www.nomadproject.io/guides/cluster/requirements.html):

> Nomad servers may need to be run on large machine instances. We suggest having between 4-8+ cores, 16-32 GB+ of memory, 40-80 GB+ of disk and significant network bandwidth. The core count and network recommendations are to ensure high throughput as Nomad heavily relies on network communication and as the Servers are managing all the nodes in the region and performing scheduling. The memory and disk requirements are due to the fact that Nomad stores all state in memory and will store two snapshots of this data onto disk. Thus disk should be at least 2 times the memory available to the server when deploying a high load cluster.

but that's probably for a rather large cluster. I bet we can get away with a small or medium.

Reply:
Oh damn, that's more than I expected! Okay, let's leave it as is and then see how they perform at relative scale.